### ISSUE TITLE
[claude] grid="auto" produces incorrect output when work shape does not exactly fill the device grid

### ISSUE THREAD

**Issue by brnorris03** [tenstorrent/tt-lang#541]
### Problem description

`@ttl.operation(grid="auto")` deterministically produces incorrect output whenever the work (M_BLOCKS x N_BLOCKS) does not exactly fill the device's compute grid. The kernel runs on every core of the auto-allocated grid, including cores that have no valid block to process. Those "idle" cores still execute the dm_read / dm_write / compute loops with out-of-bounds block indices, which corrupts the multicast handshake (and likely also corrupts adjacent DRAM via OOB writes).

This appears to be the underlying cause of:
- #513 -- `test_balanced_matmul` reported as flaky (output values such as `mean = -1.4e29`, consistent with uninitialized memory).
- #502 -- `singlecore_matmul.py` and `multicore_matmul.py` fail on hardware (both use `grid="auto"`).
- All three tests in `test/python/pipe/test_mcast_matmul.py` (`test_mcast_matmul`, `test_balanced_matmul`, `test_balanced_matmul_relu`) failing on hardware -- not only `test_balanced_matmul`.

The reported flakiness is consistent with OOB DRAM writes landing on different addresses across runs. On the wormhole_b0 setup tested here the failure is fully deterministic.

### Reproduction

Reproduced on wormhole_b0 (8x7 compute grid) and on blackhole (different grid; see below). Run from a built tree:

```bash
source build/env/activate
pytest test/python/pipe/test_mcast_matmul.py -v
```

All three tests fail in ~24s on wormhole. The minimal repro is M = K = N = 256:

```python
import torch, ttl, ttnn
from ttlang_test_utils import to_dram

TILE = 32
BLOCK_M = BLOCK_N = BLOCK_K = 8
BLOCK_SIZE = BLOCK_M * TILE  # 256

def make_kernel(M, K, N, grid):
    M_BLOCKS = M // BLOCK_SIZE
    N_BLOCKS = N // BLOCK_SIZE
    K_BLOCKS = K // BLOCK_SIZE

    @ttl.operation(grid=grid)
    def mm(a, w, out):
        NUM_COLS, NUM_ROWS = ttl.grid_size(dims=2)
        # ... same body as test_mcast_matmul.make_mcast_kernel ...
    return mm

device = ttnn.open_device(device_id=0)
a = to_dram(torch.randn(256, 256, dtype=torch.bfloat16) * 0.02, device)
w = to_dram(torch.randn(256, 256, dtype=torch.bfloat16) * 0.02, device)
out = to_dram(torch.zeros(256, 256, dtype=torch.bfloat16), device)
make_kernel(256, 256, 256, "auto")(a, w, out)
# PCC vs ttnn.matmul(a, w) -> ~0.89  FAIL

make_kernel(256, 256, 256, (1, 1))(a, w, out)
# PCC vs ttnn.matmul(a, w) -> 0.99998  PASS
```

A more complete sweep is at `repro_mcast_matmul.py` in this repo (driver script). It uses the same kernel body as `test/python/pipe/test_mcast_matmul.py::make_mcast_kernel`, and varies (M, K, N, grid).

### Sweep results (wormhole_b0, 8x7 grid)

| Config | grid="auto" | explicit grid |
|--------|-------------|---------------|
| M=K=N=256 (1x1x1) | FAIL PCC 0.89 | PASS 0.99998 (grid=(1,1)) |
| M=K=N=512 (2x2x2) | FAIL PCC 0.55 | PASS 0.99997 (grid=(2,2)) |
| M=K=N=1024 (4x4x4) | FAIL PCC 0.39 | PASS 0.99993 (grid=(4,4)) |
| M=1792 K=256 N=2048 (7x1x8 -- exactly fills 8x7) | PASS 0.99994 | PASS 0.99994 |
| M=10240 K=8192 N=13312 (baseline, 5x4 active out of 8x7) | FAIL PCC 0.54 | -- |

### Sweep results (blackhole)

The same sweep on blackhole shows the failure tracks the device grid rather than any particular configuration. Configurations that pass on wormhole now fail (and vice versa), because the alignment between the work shape and the device grid changes:

| Config | grid="auto" | explicit grid |
|--------|-------------|---------------|
| M=K=N=256 (1x1x1) | FAIL PCC 0.62 | PASS 0.99998 (grid=(1,1)) |
| M=K=N=512 (2x2x2) | FAIL PCC 0.17 | PASS 0.99997 (grid=(2,2)) |
| M=K=N=1024 (4x4x4) | FAIL PCC 0.06 | PASS 0.99995 (grid=(4,4)) |
| M=1792 K=256 N=2048 (7x1x8) | FAIL PCC 0.41 | PASS 0.99994 (grid=(8,7)) |
| M=10240 K=8192 N=13312 (baseline) | PASS 1.00 | -- |

The pattern: `grid="auto"` is correct only when M_BLOCKS x N_BLOCKS evenly fills the device grid (no core is left without a valid block). Configurations that satisfy this on one device may fail on another with a different grid shape.

For the baseline:
- `_even_split(40, NUM_ROWS=7)` returns `bpn=8`, `num_rows_used = 40/8 = 5` -> 2 idle rows
- `_even_split(52, NUM_COLS=8)` returns `bpn=13`, `num_cols_used = 52/13 = 4` -> 4 idle cols
- 36 of 56 cores are "idle" but still execute the kernel with OOB block indices.

### Root cause

In `make_mcast_kernel` (and equivalent kernels), the loop bounds use `m_blocks_per_node` / `n_blocks_per_node` from `_even_split`, but no per-core gating ensures that `node_m * m_blocks_per_node + local_mb < M_BLOCKS`. Idle cores compute mb >= M_BLOCKS / nb >= N_BLOCKS, then:
- read `a[mr:mr+BLOCK_M, kc:kc+BLOCK_K]` with `mr` past the end of A,
- read `w[kc:kc+BLOCK_K, nc:nc+BLOCK_N]` with `nc` past the end of W,
- broadcast that out-of-bounds data on the row/column multicast pipe (every multicast pipe is sized to `slice(0, NUM_COLS)` / `slice(0, NUM_ROWS)` regardless of how many cores actually have work),
- write `out[mr:mr+BLOCK_M, nc:nc+BLOCK_N]` past the end of the output.

Either the OOB reads/writes are the failure mechanism, or the multicast handshake (which expects all destinations in the configured slice to participate) is being violated when an active col multicasts B while an idle row multicasts out-of-bounds data on its own pipe. Both effects are plausible given the corruption pattern.

### Possible fixes

1. **Tighten the auto grid resolution.** `grid="auto"` should pick the smallest subgrid that holds the work, not the device's full compute grid. For these tests, a `(N_BLOCKS, M_BLOCKS)` capped by the device size would eliminate the issue.
2. **Per-core block-validity gating.** ttl could emit a guard `if mb < M_BLOCKS and nb < N_BLOCKS` around the dm/compute loops so idle cores no-op.
3. **Mcast pipe destination resolution.** `slice(0, NUM_COLS)` could resolve to "active cols only" rather than "all device cols," so the handshake is sized correctly.

(1) is the cleanest user-facing fix; (2) is a workaround that callers can apply today.

### Workaround

Pass an explicit grid sized to the work:

```python
@ttl.operation(grid=(N_BLOCKS, M_BLOCKS))   # or capped to device grid
def my_kernel(a, w, out): ...
```

This makes all listed configs pass on hardware.

### Checklist

- [ ] Decide between auto-grid tightening vs per-core gating vs both.
- [ ] Add a regression test exercising small-dim `grid="auto"` (e.g., the M=K=N=256 case above).
- [ ] Update / consolidate #513 and #502 once root cause is fixed.

**Comment by brnorris03** [comment]
Root cause is in the user-side test kernels, not the compiler. The
`make_*_kernel` factories in `test/python/pipe/test_mcast_matmul.py`
partitioned work via `_even_split` inside the closure but constructed
multicast pipes against the full device extent (`slice(0, NUM_COLS)`,
`range(NUM_ROWS)`). With `grid="auto"` resolving to the device extent,
idle nodes still executed the kernel with out-of-bounds block indices,
corrupting DRAM and the multicast handshake.

Fix in https://github.com/tenstorrent/tt-lang/pull/544:
- Migrate `make_mcast_kernel` / `make_balanced_kernel` /
  `make_balanced_relu_kernel` to `grid=callable`. The callable runs
  host-side, applies `even_split` against the device, and returns the
  active subgrid. Inside the closure, `ttl.grid_size(dims=2)` then
  yields the active extent and the existing pipe / iteration code is
  correct unchanged.
- Add a compile-time guard in `_compile_kernel`: `grid="auto"` plus any
  multicast pipe raises `ValueError` with a migration message pointing
  to the callable form. Mirror guard in `python/sim/operation.py` for
  simulator parity.
- Centralize `even_split` in `test/ttlang_test_utils.py`.
- New regression test `test/python/pipe/test_mcast_matmul_subgrid.py`
  covers the four shape regimes from the sweep, all three multicast
  variants, bf16 + fp32.
- Migration audit: `test/python/pipe/test_scatter_auto.py`,
  `examples/eltwise_pipe.py`, `examples/matmul_1d_mcast.py`,
  `examples/metal_examples/1d_mcast_matmul/ttlang/1d_tt_lang.py`.

`repro_mcast_matmul.py` is preserved with a header comment describing
the bug pattern; the canonical fixed kernel is in
`test/python/pipe/test_mcast_matmul.py`.

Hardware verification on wormhole_b0: all three previously failing
`test_mcast_matmul.py` tests pass at PCC ≥ 0.99, the new subgrid
regression passes across all parameterizations, and the negative test
confirms `grid="auto"` + multicast pipes raises with the migration
message.
